File: AdMob/iOS/AppLovinCustomEventBanner.m
//
//  AppLovinCustomEventBanner.m
//
//
//  Created by Thomas So on 4/12/17.
//
//

#import "AppLovinCustomEventBanner.h"

#if __has_include(<AppLovinSDK/AppLovinSDK.h>)
    #import <AppLovinSDK/AppLovinSDK.h>
#else
    #import "ALAdView.h"
#endif

// Convenience macro for checking if AppLovin SDK has support for zones
#define HAS_ZONES_SUPPORT [[ALSdk shared].adService respondsToSelector: @selector(loadNextAdForZoneIdentifier:andNotify:)]
#define DEFAULT_ZONE @""

@interface ALAdMobAdView : ALAdView @end

/**
 * The receiver object of the ALAdView's and ALAdService's delegates. This is used to prevent a retain cycle between the ALAdView and AppLovinBannerCustomEvent.
 */
@interface AppLovinAdMobBannerDelegate : NSObject<ALAdDisplayDelegate>
@property (nonatomic, weak) AppLovinCustomEventBanner *parentCustomEvent;
- (instancetype)initWithCustomEvent:(AppLovinCustomEventBanner *)parentCustomEvent;
@end

@interface AppLovinCustomEventBanner()<ALAdLoadDelegate>
@property (nonatomic, strong) ALAdView *adView;
@property (nonatomic,   copy) NSString *zoneIdentifier; // The zone identifier this instance of the custom event is loading for
@end

@implementation AppLovinCustomEventBanner
@synthesize delegate;

static const BOOL kALLoggingEnabled = YES;
static NSString *const kALAdMobMediationErrorDomain = @"com.applovin.sdk.mediation.admob.errorDomain";

static const CGFloat kALBannerHeightOffsetTolerance = 10.0f;
static const CGFloat kALBannerStandardHeight = 50.0f;

// A dictionary of Zone -> AdView to be shared by instances of the custom event to prevent redundant recreation of our `ALAdView`s.
static NSMutableDictionary<NSString *, ALAdView *> *ALGlobalAdViews;

// A dictionary of Zone -> Queue of `ALAd`s to be shared by instances of the custom event.
// This prevents skipping of ads as this adapter will be re-created and preloaded
// on every ad load regardless if ad was actually displayed or not.
static NSMutableDictionary<NSString *, NSMutableArray<ALAd *> *> *ALGlobalAdViewAds;
static NSObject *ALGlobalAdViewAdsLock;

#pragma mark - Class Initialization

+ (void)initialize
{
    [super initialize];
    
    ALGlobalAdViews = [NSMutableDictionary dictionary];
    
    ALGlobalAdViewAds = [NSMutableDictionary dictionary];
    ALGlobalAdViewAdsLock = [[NSObject alloc] init];
}

#pragma mark - GADCustomEventBanner Protocol

- (void)requestBannerAd:(GADAdSize)adSize parameter:(NSString *)serverParameter label:(NSString *)serverLabel request:(GADCustomEventRequest *)request
{
    [self log: @"Requesting AppLovin banner of size %@", NSStringFromGADAdSize(adSize)];
    
    // Convert requested size to AppLovin Ad Size
    ALAdSize *appLovinAdSize = [self appLovinAdSizeFromRequestedSize: adSize];
    if ( appLovinAdSize )
    {
        [[ALSdk shared] setPluginVersion: @"AdMob-2.3"];
        
        CGSize size = CGSizeFromGADAdSize(adSize);
        
        // Zones support is available on AppLovin SDK 4.5.0 and higher
        if ( HAS_ZONES_SUPPORT && request.additionalParameters[@"zone_id"] )
        {
            self.zoneIdentifier = request.additionalParameters[@"zone_id"];
        }
        else
        {
            self.zoneIdentifier = DEFAULT_ZONE;
        }
        
        
        self.adView = ALGlobalAdViews[self.zoneIdentifier];
        
        // Check if we already have an ALAdView for the given zone
        if ( !self.adView )
        {
            // If this is a default Zone, create the incentivized ad normally
            if ( [DEFAULT_ZONE isEqualToString: self.zoneIdentifier] )
            {
                self.adView = [[ALAdMobAdView alloc] initWithFrame: CGRectMake(0.0f, 0.0f, size.width, size.height)
                                                              size: appLovinAdSize
                                                               sdk: [ALSdk shared]];
            }
            // Otherwise, use the Zones API
            else
            {
                self.adView = [self adViewWithAdSize: appLovinAdSize zoneIdentifier: self.zoneIdentifier];
            }
            
            ALGlobalAdViews[self.zoneIdentifier] = self.adView;
        }
        
        
        AppLovinAdMobBannerDelegate *delegate = [[AppLovinAdMobBannerDelegate alloc] initWithCustomEvent: self];
        self.adView.adDisplayDelegate = delegate;
        
        // Already have a preloaded ad
        if ( [[self class] hasEnqueuedAdForZoneIdentifier: self.zoneIdentifier] )
        {
            [self.delegate customEventBanner: self didReceiveAd: self.adView];
        }
        else
        {
            // If this is a default Zone, create the incentivized ad normally
            if ( [DEFAULT_ZONE isEqualToString: self.zoneIdentifier] )
            {
                [[ALSdk shared].adService loadNextAd: appLovinAdSize andNotify: self];
            }
            // Otherwise, use the Zones API
            else
            {
                // Dynamically load an ad for a given zone without breaking backwards compatibility for publishers on older SDKs
                [[ALSdk shared].adService performSelector: @selector(loadNextAdForZoneIdentifier:andNotify:)
                                               withObject: self.zoneIdentifier
                                               withObject: self];
            }
        }
    }
    else
    {
        [self log: @"Failed to create an AppLovin Banner with invalid size"];
        
        NSError *error = [NSError errorWithDomain: kALAdMobMediationErrorDomain
                                             code: kGADErrorMediationInvalidAdSize
                                         userInfo: nil];
        [self.delegate customEventBanner: self didFailAd: error];
    }
}


#pragma mark - AppLovin Ad Load Delegate

- (void)adService:(ALAdService *)adService didLoadAd:(ALAd *)ad
{
    [self log: @"Banner did load ad: %@", ad.adIdNumber];
    
    if ( !self.adView.window )
    {
        [AppLovinCustomEventBanner enqueueAd: ad forZoneIdentifier: self.zoneIdentifier];
    }
    else
    {
        // Check if we have enqueued ad already
        ALAd *enqueuedAd = [AppLovinCustomEventBanner dequeueAdForZoneIdentifier: self.zoneIdentifier];
        if ( enqueuedAd )
        {
            [self.adView render: enqueuedAd];
            [AppLovinCustomEventBanner enqueueAd: ad forZoneIdentifier: self.zoneIdentifier];
        }
        // No enqueued ad, render newly loaded ad
        else
        {
            [self.adView render: ad];
        }
    }
    
    [self.delegate customEventBanner: self didReceiveAd: self.adView];
}

- (void)adService:(ALAdService *)adService didFailToLoadAdWithError:(int)code
{
    [self log: @"Banner failed to load with error: %d", code];
    
    NSError *error = [NSError errorWithDomain: kALAdMobMediationErrorDomain
                                         code: [self toAdMobErrorCode: code]
                                     userInfo: nil];
    
    // If CURRENT ad request was a no fill, check against enqueued ads
    if ( code == kALErrorCodeNoFill )
    {
        ALAd *preloadedAd = [AppLovinCustomEventBanner dequeueAdForZoneIdentifier: self.zoneIdentifier];
        
        // There is an enqueued ad, use that
        if ( preloadedAd )
        {
            [self log: @"Using enqueued ad instead..."];
            [self adService: adService didLoadAd: preloadedAd];
        }
        else
        {
            [self.delegate customEventBanner: self didFailAd: error];
        }
    }
    else
    {
        [self.delegate customEventBanner: self didFailAd: error];
    }
}

#pragma mark - Utility Methods

+ (BOOL)hasEnqueuedAdForZoneIdentifier:(NSString *)zoneIdentifier
{
    @synchronized ( ALGlobalAdViewAdsLock )
    {
        return ALGlobalAdViewAds[zoneIdentifier].count > 0;
    }
}

+ (alnullable ALAd *)dequeueAdForZoneIdentifier:(NSString *)zoneIdentifier
{
    @synchronized ( ALGlobalAdViewAdsLock )
    {
        ALAd *preloadedAd;
        
        NSMutableArray<ALAd *> *preloadedAds = ALGlobalAdViewAds[zoneIdentifier];
        if ( preloadedAds.count > 0 )
        {
            preloadedAd = preloadedAds[0];
            [preloadedAds removeObjectAtIndex: 0];
        }
        
        return preloadedAd;
    }
}

+ (void)enqueueAd:(ALAd *)ad forZoneIdentifier:(NSString *)zoneIdentifier
{
    @synchronized ( ALGlobalAdViewAdsLock )
    {
        NSMutableArray<ALAd *> *preloadedAds = ALGlobalAdViewAds[zoneIdentifier];
        if ( !preloadedAds )
        {
            preloadedAds = [NSMutableArray array];
            ALGlobalAdViewAds[zoneIdentifier] = preloadedAds;
        }
        
        [preloadedAds addObject: ad];
    }
}

/**
 * Dynamically create an instance of ALAdView with a given zone without breaking backwards compatibility for publishers on older SDKs.
 */
- (ALAdView *)adViewWithAdSize:(ALAdSize *)adSize zoneIdentifier:(NSString *)zoneIdentifier
{
    // Prematurely create instance of ALAdView to store initialized one in later
    ALAdView *adView = [ALAdMobAdView alloc];
    
    // We must use NSInvocation over performSelector: for initializers
    NSMethodSignature *methodSignature = [ALAdView instanceMethodSignatureForSelector: @selector(initWithSize:zoneIdentifier:)];
    NSInvocation *inv = [NSInvocation invocationWithMethodSignature: methodSignature];
    [inv setSelector: @selector(initWithSize:zoneIdentifier:)];
    [inv setArgument: &adSize atIndex: 2];
    [inv setArgument: &zoneIdentifier atIndex: 3];
    [inv setReturnValue: &adView];
    [inv invokeWithTarget: adView];
    
    return adView;
}

- (ALAdSize *)appLovinAdSizeFromRequestedSize:(GADAdSize)size
{
    if ( GADAdSizeEqualToSize(kGADAdSizeBanner, size ) || GADAdSizeEqualToSize(kGADAdSizeLargeBanner, size ) )
    {
        return [ALAdSize sizeBanner];
    }
    else if ( GADAdSizeEqualToSize(kGADAdSizeMediumRectangle, size) )
    {
        return [ALAdSize sizeMRec];
    }
    else if ( GADAdSizeEqualToSize(kGADAdSizeLeaderboard, size) )
    {
        return [ALAdSize sizeLeader];
    }
    // This is not a one of AdMob's predefined size
    else
    {
        CGSize frameSize = size.size;
        
        // Attempt to check for fluid size
        if ( CGRectGetWidth([UIScreen mainScreen].bounds) == frameSize.width )
        {
            CGFloat frameHeight = frameSize.height;
            if ( frameHeight == CGSizeFromGADAdSize(kGADAdSizeBanner).height || frameHeight == CGSizeFromGADAdSize(kGADAdSizeLargeBanner).height )
            {
                return [ALAdSize sizeBanner];
            }
            else if ( frameHeight == CGSizeFromGADAdSize(kGADAdSizeMediumRectangle).height )
            {
                return [ALAdSize sizeMRec];
            }
            else if ( frameHeight == CGSizeFromGADAdSize(kGADAdSizeLeaderboard).height )
            {
                return [ALAdSize sizeLeader];
            }
        }
        
        // Assume fluid width, and check for height with offset tolerance
        CGFloat offset = ABS(kALBannerStandardHeight - frameSize.height);
        if ( offset <= kALBannerHeightOffsetTolerance )
        {
            return [ALAdSize sizeBanner];
        }
    }
    
    [self log: @"Unable to retrieve AppLovin size from GADAdSize: %@", NSStringFromGADAdSize(size)];
    
    return nil;
}

- (void)log:(NSString *)format, ...
{
    if ( kALLoggingEnabled )
    {
        va_list valist;
        va_start(valist, format);
        NSString *message = [[NSString alloc] initWithFormat: format arguments: valist];
        va_end(valist);
        
        NSLog(@"AppLovinCustomEventBanner: %@", message);
    }
}

- (GADErrorCode)toAdMobErrorCode:(int)appLovinErrorCode
{
    if ( appLovinErrorCode == kALErrorCodeNoFill )
    {
        return kGADErrorMediationNoFill;
    }
    else if ( appLovinErrorCode == kALErrorCodeAdRequestNetworkTimeout )
    {
        return kGADErrorTimeout;
    }
    else if ( appLovinErrorCode == kALErrorCodeInvalidResponse )
    {
        return kGADErrorReceivedInvalidResponse;
    }
    else if ( appLovinErrorCode == kALErrorCodeUnableToRenderAd )
    {
        return kGADErrorServerError;
    }
    else
    {
        return kGADErrorInternalError;
    }
}

@end

@implementation AppLovinAdMobBannerDelegate

#pragma mark - Initialization

- (instancetype)initWithCustomEvent:(AppLovinCustomEventBanner *)parentCustomEvent
{
    self = [super init];
    if ( self )
    {
        self.parentCustomEvent = parentCustomEvent;
    }
    return self;
}

#pragma mark - Ad Display Delegate

- (void)ad:(ALAd *)ad wasDisplayedIn:(UIView *)view
{
    [self.parentCustomEvent log: @"Banner displayed"];
}

- (void)ad:(ALAd *)ad wasHiddenIn:(UIView *)view
{
    [self.parentCustomEvent log: @"Banner dismissed"];
}

- (void)ad:(ALAd *)ad wasClickedIn:(UIView *)view
{
    [self.parentCustomEvent log: @"Banner clicked"];
    
    [self.parentCustomEvent.delegate customEventBannerWasClicked: self.parentCustomEvent];
    [self.parentCustomEvent.delegate customEventBannerWillLeaveApplication: self.parentCustomEvent];
}

@end

@implementation ALAdMobAdView @end

/**
 * This category provides a way to have an `ALAdView` to dynamically render an enqueued ad WHEN needed.
 */
@interface ALAdMobAdView (AdMob)
@property (nonatomic, copy, readonly) NSString *zoneIdentifier;
@end

@implementation ALAdMobAdView (AdMob)
@dynamic zoneIdentifier;

- (void)didMoveToWindow
{
    [super didMoveToWindow];
    
    if ( self.window )
    {
        NSString *zoneIdentifier = ( HAS_ZONES_SUPPORT && self.zoneIdentifier ) ? self.zoneIdentifier : DEFAULT_ZONE;
        
        ALAd *preloadedAd = [AppLovinCustomEventBanner dequeueAdForZoneIdentifier: zoneIdentifier];
        if ( preloadedAd )
        {
            [self render: preloadedAd];
        }
        // Something is wrong... no preloaded ad provided... manually load an ad if none provided
        else
        {
            [self loadNextAd];
        }
    }
}

@end

File: AdMob/Android/AppLovinCustomEventBanner.java
package AdMob.Android;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.applovin.adview.AppLovinAdView;
import com.applovin.sdk.AppLovinAd;
import com.applovin.sdk.AppLovinAdClickListener;
import com.applovin.sdk.AppLovinAdDisplayListener;
import com.applovin.sdk.AppLovinAdLoadListener;
import com.applovin.sdk.AppLovinAdSize;
import com.applovin.sdk.AppLovinErrorCodes;
import com.applovin.sdk.AppLovinSdk;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.mediation.MediationAdRequest;
import com.google.android.gms.ads.mediation.customevent.CustomEventBanner;
import com.google.android.gms.ads.mediation.customevent.CustomEventBannerListener;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;

import static android.util.Log.DEBUG;
import static android.util.Log.ERROR;

/**
 * AppLovin SDK banner adapter for AdMob.
 * <p>
 * Created by thomasso on 4/12/17.
 */

public class AppLovinCustomEventBanner
        implements CustomEventBanner, AppLovinAdLoadListener
{
    private static final boolean LOGGING_ENABLED = true;
    private static final Handler UI_HANDLER      = new Handler( Looper.getMainLooper() );
    private static final String  DEFAULT_ZONE    = "";

    private static final int BANNER_STANDARD_HEIGHT         = 50;
    private static final int BANNER_HEIGHT_OFFSET_TOLERANCE = 10;

    // A dictionary of Zone -> Queue of `AppLovinAd`s to be shared by instances of the custom event.
    // This prevents skipping of ads as this adapter will be re-created and preloaded
    // on every ad load regardless if ad was actually displayed or not.
    private static final Map<String, Queue<AppLovinAd>> GLOBAL_AD_VIEW_ADS      = new HashMap<String, Queue<AppLovinAd>>();
    private static final Object                         GLOBAL_AD_VIEW_ADS_LOCK = new Object();

    private CustomEventBannerListener customEventBannerListener;

    private AppLovinAdView adView;
    private String         zoneId; // The zone identifier this instance of the custom event is loading for

    //
    // AdMob Custom Event Methods
    //

    @Override
    public void requestBannerAd(final Context context, final CustomEventBannerListener customEventBannerListener, final String serverParameter, final AdSize adSize, final MediationAdRequest mediationAdRequest, final Bundle customEventExtras)
    {
        this.customEventBannerListener = customEventBannerListener;

        // SDK versions BELOW 7.1.0 require a instance of an Activity to be passed in as the context
        if ( AppLovinSdk.VERSION_CODE < 710 && !( context instanceof Activity ) )
        {
            log( ERROR, "Unable to request AppLovin banner. Invalid context provided." );
            customEventBannerListener.onAdFailedToLoad( AdRequest.ERROR_CODE_INTERNAL_ERROR );

            return;
        }

        log( DEBUG, "Requesting AppLovin banner of size: " + adSize );

        final AppLovinAdSize appLovinAdSize = appLovinAdSizeFromAdMobAdSize( adSize );
        if ( appLovinAdSize != null )
        {
            final AppLovinSdk sdk = AppLovinSdk.getInstance( context );
            sdk.setPluginVersion( "AdMob-2.1" );

            // Zones support is available on AppLovin SDK 7.5.0 and higher
            if ( AppLovinSdk.VERSION_CODE >= 750 && customEventExtras != null && customEventExtras.containsKey( "zone_id" ) )
            {
                zoneId = customEventExtras.getString( "zone_id" );
            }
            else
            {
                zoneId = DEFAULT_ZONE;
            }

            adView = createAdView( zoneId, appLovinAdSize, context, customEventBannerListener );

            // Already have a preloaded ad
            if ( hasEnqueuedAd( zoneId ) )
            {
                log( DEBUG, "Found enqueued ad while requesting for banner. Using that..." );
                customEventBannerListener.onAdLoaded( adView );
            }
            else
            {
                // If this is a default Zone, load the ad normally
                if ( DEFAULT_ZONE.equals( zoneId ) )
                {
                    log( DEBUG, "Loading next ad for default zone" );
                    AppLovinSdk.getInstance( context ).getAdService().loadNextAd( appLovinAdSize, this );
                }
                // Otherwise, use the Zones API
                else
                {
                    // Dynamically load an ad for a given zone without breaking backwards compatibility for publishers on older SDKs
                    try
                    {
                        log( DEBUG, "Loading next ad for custom zone: " + zoneId );
                        final Method method = sdk.getAdService().getClass().getMethod( "loadNextAdForZoneId", String.class, AppLovinAdLoadListener.class );
                        method.invoke( sdk.getAdService(), zoneId, this );
                    }
                    catch ( Throwable th )
                    {
                        log( ERROR, "Unable to load ad for zone: " + zoneId + "..." );
                        customEventBannerListener.onAdFailedToLoad( AdRequest.ERROR_CODE_INVALID_REQUEST );
                    }
                }
            }
        }
        else
        {
            log( ERROR, "Unable to request AppLovin banner" );
            customEventBannerListener.onAdFailedToLoad( AdRequest.ERROR_CODE_INTERNAL_ERROR );
        }
    }

    @Override
    public void onDestroy() {}

    @Override
    public void onPause()
    {
        log( DEBUG, "onPause() called" );
        if ( adView != null ) adView.pause();
    }

    @Override
    public void onResume()
    {
        log( DEBUG, "onResume() called" );
        if ( adView != null ) adView.resume();
    }

    //
    // Ad Load Listener
    //

    @Override
    public void adReceived(final AppLovinAd ad)
    {
        // AdMob requires `onAdLoaded()` and `onAdFailedToLoad()` callbacks to be called on main queue
        UI_HANDLER.post( new Runnable()
        {
            @Override
            public void run()
            {
                log( DEBUG, "Successfully loaded banner ad" );

                if ( !adView.isAttachedToWindow() )
                {
                    log( DEBUG, "Enqueuing loaded ad..." );
                    enqueueAd( ad, zoneId );
                }
                else
                {
                    log( DEBUG, "Rendering loaded ad..." );
                    adView.renderAd( ad );
                }

                customEventBannerListener.onAdLoaded( adView );
            }
        } );
    }

    @Override
    public void failedToReceiveAd(final int errorCode)
    {
        // AdMob requires `onAdLoaded()` and `onAdFailedToLoad()` callbacks to be called on main queue
        UI_HANDLER.post( new Runnable()
        {
            @Override
            public void run()
            {
                log( ERROR, "Failed to load banner ad with code: " + errorCode );

                // If CURRENT ad request was a no fill, check against enqueued ads
                if ( errorCode == AppLovinErrorCodes.NO_FILL )
                {
                    final AppLovinAd preloadedAd = dequeueAd( zoneId );

                    // There is an enqueued ad, use that
                    if ( preloadedAd != null )
                    {
                        log( DEBUG, "Using enqueued ad instead..." );
                        adReceived( preloadedAd );

                        return;
                    }
                }

                customEventBannerListener.onAdFailedToLoad( toAdMobErrorCode( errorCode ) );
            }
        } );
    }

    //
    // Utility Methods
    //

    private static boolean hasEnqueuedAd(final String zoneId)
    {
        synchronized ( GLOBAL_AD_VIEW_ADS_LOCK )
        {
            final Queue<AppLovinAd> preloadedAds = GLOBAL_AD_VIEW_ADS.get( zoneId );
            return preloadedAds != null && !preloadedAds.isEmpty();
        }
    }

    private static AppLovinAd dequeueAd(final String zoneId)
    {
        synchronized ( GLOBAL_AD_VIEW_ADS_LOCK )
        {
            AppLovinAd preloadedAd = null;

            final Queue<AppLovinAd> preloadedAds = GLOBAL_AD_VIEW_ADS.get( zoneId );
            if ( preloadedAds != null && !preloadedAds.isEmpty() )
            {
                preloadedAd = preloadedAds.poll();
            }

            log( DEBUG, "Dequeued ad: " + preloadedAd + " for zone: " + zoneId );

            return preloadedAd;
        }
    }

    private static void enqueueAd(final AppLovinAd ad, final String zoneId)
    {
        synchronized ( GLOBAL_AD_VIEW_ADS_LOCK )
        {
            Queue<AppLovinAd> preloadedAds = GLOBAL_AD_VIEW_ADS.get( zoneId );
            if ( preloadedAds == null )
            {
                preloadedAds = new LinkedList<AppLovinAd>();
                GLOBAL_AD_VIEW_ADS.put( zoneId, preloadedAds );
            }

            preloadedAds.offer( ad );

            log( DEBUG, "Enqueued ad: " + ad + " for zone: " + zoneId + ". Size: " + preloadedAds.size() );
        }
    }

    private AppLovinAdView createAdView(final String zoneId, final AppLovinAdSize size, final Context parentContext, final CustomEventBannerListener customEventBannerListener)
    {
        AppLovinAdMobAdView adView = null;

        try
        {
            // AppLovin SDK < 7.1.0 uses an Activity, as opposed to Context in >= 7.1.0
            final Class<?> contextClass = ( AppLovinSdk.VERSION_CODE < 710 ) ? Activity.class : Context.class;

            final Constructor<?> constructor;

            // If this is a default Zone, create the incentivized ad normally
            if ( DEFAULT_ZONE.equals( zoneId ) )
            {
                adView = new AppLovinAdMobAdView( size, parentContext );
            }
            // Otherwise, use the Zones API
            else
            {
                // Dynamically create an instance of AppLovinAdView with a given zone without breaking backwards compatibility for publishers on older SDKs.
                constructor = AppLovinAdMobAdView.class.getConstructor( getClass(), AppLovinAdSize.class, String.class, contextClass );
                adView = (AppLovinAdMobAdView) constructor.newInstance( this, size, zoneId, parentContext );
            }

            final AppLovinAdMobBannerListener listener = new AppLovinAdMobBannerListener();
            adView.setAdDisplayListener( listener );
            adView.setAdClickListener( listener );
        }
        catch ( Throwable th )
        {
            log( ERROR, "Unable to create AppLovinAdView. Because of: " + Log.getStackTraceString( th ) );
            customEventBannerListener.onAdFailedToLoad( AdRequest.ERROR_CODE_INTERNAL_ERROR );
        }

        return adView;
    }

    private AppLovinAdSize appLovinAdSizeFromAdMobAdSize(final AdSize adSize)
    {
        if ( AdSize.BANNER.equals( adSize ) || AdSize.LARGE_BANNER.equals( adSize ) )
        {
            return AppLovinAdSize.BANNER;
        }
        else if ( AdSize.MEDIUM_RECTANGLE.equals( adSize ) )
        {
            return AppLovinAdSize.MREC;
        }
        else if ( AdSize.LEADERBOARD.equals( adSize ) )
        {
            return AppLovinAdSize.LEADER;
        }
        // This is not a one of AdMob's predefined size
        else
        {
            // Assume fluid width, and check for height with offset tolerance
            final int offset = Math.abs( BANNER_STANDARD_HEIGHT - adSize.getHeight() );
            if ( offset <= BANNER_HEIGHT_OFFSET_TOLERANCE )
            {
                return AppLovinAdSize.BANNER;
            }
        }

        return null;
    }

    private static void log(final int priority, final String message)
    {
        if ( LOGGING_ENABLED )
        {
            Log.println( priority, "AppLovinBanner", message );
        }
    }

    private static int toAdMobErrorCode(final int applovinErrorCode)
    {
        if ( applovinErrorCode == AppLovinErrorCodes.NO_FILL )
        {
            return AdRequest.ERROR_CODE_NO_FILL;
        }
        else if ( applovinErrorCode == AppLovinErrorCodes.NO_NETWORK || applovinErrorCode == AppLovinErrorCodes.FETCH_AD_TIMEOUT )
        {
            return AdRequest.ERROR_CODE_NETWORK_ERROR;
        }
        else
        {
            return AdRequest.ERROR_CODE_INTERNAL_ERROR;
        }
    }

    /**
     * The receiver object of the AppLovinAdView's and AppLovinAdService's listeners.
     */
    private class AppLovinAdMobBannerListener
            implements AppLovinAdDisplayListener, AppLovinAdClickListener
    {
        @Override
        public void adDisplayed(final AppLovinAd ad)
        {
            log( DEBUG, "Banner displayed" );
        }

        @Override
        public void adHidden(final AppLovinAd ad)
        {
            log( DEBUG, "Banner dismissed" );
        }

        @Override
        public void adClicked(final AppLovinAd ad)
        {
            log( DEBUG, "Banner clicked" );

            customEventBannerListener.onAdOpened();
            customEventBannerListener.onAdLeftApplication();
        }
    }

    /**
     * This subclass provides a way to have an `AppLovinAdView` to dynamically render an enqueued ad WHEN needed.
     */
    public class AppLovinAdMobAdView
            extends AppLovinAdView
    {
        private AppLovinAdMobAdView(final AppLovinAdSize adSize, final Context context)
        {
            this( adSize, DEFAULT_ZONE, context );
        }

        public AppLovinAdMobAdView(final AppLovinAdSize adSize, final String zoneId, final Context context)
        {
            // TODO: Make dynamic
            super( adSize, zoneId, context );

            // To prevent weird AdMob case where it quickly sets visibility to GONE _after_ attaching to window and displaying ad, causing us to render an empty ad
            setAutoDestroy( false );
        }

        @Override
        protected void onAttachedToWindow()
        {
            super.onAttachedToWindow();

            log( DEBUG, "AppLovinAdView attached to window" );

            final AppLovinAd preloadedAd = dequeueAd( zoneId );
            if ( preloadedAd != null )
            {
                log( DEBUG, "Rendering enqueued ad..." );
                renderAd( preloadedAd );
            }
            // Something is wrong... no preloaded ad provided... manually load an ad if none provided
            else
            {
                log( DEBUG, "Loading new ad..." );
                loadNextAd();
            }
        }
    }
}
